src/Compilers/Core/Portable/Syntax/SyntaxDiffer.cs

```diff
@@ -593,37 +638,73 @@ private static TextSpan GetSpan(Stack<SyntaxNodeOrToken> stack, int first, int l
             return TextSpan.FromBounds(start, end);
         }
 
-        private static Stack<SyntaxNodeOrToken> Combine(Stack<SyntaxNodeOrToken> first, Stack<SyntaxNodeOrToken> next)
+        private static TextSpan GetSpan(Queue<SyntaxNodeOrToken> queue, int first, int length)
+        {
+            int start = -1, end = -1, i = 0;
+            foreach (var n in queue)
+            {
+                if (i == first)
+                {
+                    start = n.Position;
+                }
+
+                if (i == first + length - 1)
+                {
+                    end = n.EndPosition;
+                    break;
+                }
+
+                i++;
+            }
+
+            Debug.Assert(start >= 0);
+            Debug.Assert(end >= 0);
+
+            return TextSpan.FromBounds(start, end);
+        }
+
+        private static Queue<SyntaxNodeOrToken> Combine(Queue<SyntaxNodeOrToken> first, Queue<SyntaxNodeOrToken> next)
         {
-            if (first == null)
+            if (first == null || first.Count == 0)
             {
                 return next;
             }
 
-            if (next == null)
+            if (next == null || next.Count == 0)
             {
                 return first;
             }
 
-            var nodes = ToArray(first, first.Count);
-            for (int i = 0; i < nodes.Length; i++)
+            foreach (var nodeOrToken in next)
             {
-                next.Push(nodes[i]);
+                first.Enqueue(nodeOrToken);
```

Comment:
oh wow, this mutates in place...    I feel like i need a good explanation on how this whole algorithm works, and how the intermediary data structures are manipulated.  (this is not for you @sharwell , it's just a general issue with this overall component).